File: backend/app/schemas/what_if.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class WhatIfScenarioRequest(BaseModel):
    scenario_name: Optional[str] = Field(default=None, description="Optional label for the scenario")
    modifications: Dict[str, Any] = Field(default_factory=dict, description="Feature overrides for simulation")


class WhatIfFeatureChange(BaseModel):
    feature: str
    label: str
    baseline_value: Any = None
    scenario_value: Any = None
    unit: Optional[str] = None


class WhatIfFeatureField(BaseModel):
    field: str
    label: str
    unit: Optional[str] = None
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    input_type: str = "number"
    baseline_value: Any = None


class WhatIfSnapshot(BaseModel):
    risk_score: float
    risk_label: str
    features: Dict[str, Any] = Field(default_factory=dict)
    top_features: List[Dict[str, Any]] = Field(default_factory=list)


class WhatIfDelta(BaseModel):
    absolute: float
    relative_percent: float
    direction: str


class WhatIfAnalysis(BaseModel):
    summary: str = ""
    drivers: List[str] = Field(default_factory=list)
    clinical_interpretation: str = ""
    cautions: List[str] = Field(default_factory=list)
    chatbot_summary: Optional[str] = None


class WhatIfBaselineResponse(BaseModel):
    patient_id: str
    baseline: WhatIfSnapshot
    modifiable_fields: List[WhatIfFeatureField] = Field(default_factory=list)
    generated_at: str


class WhatIfCompareResponse(BaseModel):
    patient_id: str
    scenario_name: Optional[str] = None
    baseline: WhatIfSnapshot
    scenario: WhatIfSnapshot
    changes: List[WhatIfFeatureChange] = Field(default_factory=list)
    risk_delta: WhatIfDelta
    analysis: WhatIfAnalysis
    requires_clinician_review: bool = True

File: frontend/services/whatIf.ts
import { apiClient } from './api';

export interface WhatIfFeatureField {
  field: string;
  label: string;
  unit?: string | null;
  min_value?: number | null;
  max_value?: number | null;
  input_type: 'number' | 'text';
  baseline_value?: string | number | null;
}

export interface WhatIfSnapshot {
  risk_score: number;
  risk_label: 'low' | 'medium' | 'high';
  features: Record<string, string | number | null>;
  top_features: Array<{
    name?: string;
    feature?: string;
    importance?: number;
    value?: string;
  }>;
}

export interface WhatIfChange {
  feature: string;
  label: string;
  baseline_value: string | number | null;
  scenario_value: string | number | null;
  unit?: string | null;
}

export interface WhatIfCompareResponse {
  patient_id: string;
  patient_name?: string;
  scenario_name?: string | null;
  baseline: WhatIfSnapshot;
  scenario: WhatIfSnapshot;
  changes: WhatIfChange[];
  risk_delta: {
    absolute: number;
    relative_percent: number;
    direction: 'decrease' | 'increase' | 'no_change';
  };
  analysis: {
    summary: string;
    drivers: string[];
    clinical_interpretation: string;
    cautions: string[];
    chatbot_summary?: string | null;
  };
  requires_clinician_review: boolean;
}

export interface WhatIfBaselineResponse {
  patient_id: string;
  patient_name?: string;
  baseline: WhatIfSnapshot;
  modifiable_fields: WhatIfFeatureField[];
  generated_at: string;
}

class WhatIfService {
  async getBaseline(patientId: string): Promise<WhatIfBaselineResponse> {
    return apiClient.get<WhatIfBaselineResponse>(`/what-if/patients/${patientId}/baseline`);
  }

  async compareRisk(
    patientId: string,
    payload: {
      scenario_name?: string;
      modifications: Record<string, unknown>;
    },
  ): Promise<WhatIfCompareResponse> {
    return apiClient.post<WhatIfCompareResponse>(
      `/what-if/patients/${patientId}/risk/compare`,
      payload,
    );
  }
}

export const whatIfService = new WhatIfService();
